### PR TITLE
fix: [SRM-16172]: Add ms-reconciliation-status API

### PR DESCRIPTION
**Pull request by naveen-harness**
## Describe your changes

Given a templateIdentifier and templateVersion, ms-reconciliation-statuses API returns all the monitored services which are created using template and their reconciliation status. 
## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

### Rebase Instructions:
To rebase your branch onto the latest changes in the target branch, simply use the following command in a comment: `/rebase`

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/55762)

### CODE REVIEW

**Comment by naveen-harness** [comment]
trigger uts

**Review by deepak-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55762)* status: 8 file(s) to review. 1 change requests. 0 of 1 approvals obtained. 

___
*[`srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/beans/monitoredService/MSReconciliationStatus.java` line 10 at r1](https://reviewable.io/reviews/harness/harness-core/55762#-Nju6rGn3Fl62tG6U2D2:-Nju6rGn3Fl62tG6U2D3:bojr0a3) ([raw file](https://github.com/harness/harness-core/blob/f2fca8b1f69c85ff87ff70e550d67b4e47266302/srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/beans/monitoredService/MSReconciliationStatus.java#L10)):*
> ```Java
> package io.harness.cvng.core.beans.monitoredService;
> 
> public enum MSReconciliationStatus {
> ```

I think ReconciliationStatus is good enough. Can be used outside MS as well.
___
*[`srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java` line 2356 at r1](https://reviewable.io/reviews/harness/harness-core/55762#-Nju72zg8kpM1G7tdDxV:-Nju72zg8kpM1G7tdDxW:bjqe0za) ([raw file](https://github.com/harness/harness-core/blob/f2fca8b1f69c85ff87ff70e550d67b4e47266302/srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java#L2356)):*
> ```Java
>             .filter(MonitoredServiceKeys.templateMetadata + ".isTemplateByReference", true)
>             .filter(MonitoredServiceKeys.templateMetadata + ".templateIdentifier", templateIdentifier)
>             .filter(MonitoredServiceKeys.templateMetadata + ".versionLabel", templateVersionLabel)
> ```

check this: https://github.com/harness/harness-core/blob/48aefdff5792df806b05647790d5bcefd147a540/srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/entities/VerificationTask.java#L116C3-L116C3
___
*[`srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java` line 2374 at r1](https://reviewable.io/reviews/harness/harness-core/55762#-Nju79OC61EWrlG-V1VO:-Nju79OC61EWrlG-V1VP:b-poaaoo) ([raw file](https://github.com/harness/harness-core/blob/f2fca8b1f69c85ff87ff70e550d67b4e47266302/srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java#L2374)):*
> ```Java
>                       .environmentIdentifier(monitoredService.getEnvironmentIdentifier())
>                       .lastReconciledTimestamp(monitoredService.getTemplateMetadata().getLastReconciliationTime());
>               if (monitoredService.getTemplateMetadata().getTemplateVersionNumber() == versionNumber) {
> ```

Can be a static function inside enum
___
*[`srm-service/modules/cv-nextgen-service/service/src/test/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImplTest.java` line 502 at r1](https://reviewable.io/reviews/harness/harness-core/55762#-Nju7GSK5p_buLjX8vfN:-Nju7GSK5p_buLjX8vfO:bjmb8pd) ([raw file](https://github.com/harness/harness-core/blob/f2fca8b1f69c85ff87ff70e550d67b4e47266302/srm-service/modules/cv-nextgen-service/service/src/test/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImplTest.java#L502)):*
> ```Java
> 
>   @Test
>   @Owner(developers = ABHIJITH)
> ```

correct username

**Review by deepak-harness — APPROVED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55762)* status: 10 file(s) to review.

**Comment by karansaraswat19** [comment]
trigger uts

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java [hunk 1]
```diff
@@ -2340,6 +2343,45 @@ public List<ActiveServiceDTO> listActiveMonitoredServices(ProjectParams projectP
...
 
+  @Override
+  public PageResponse<MonitoredServiceReference> getMonitoredServiceReconciliationStatuses(
+      ProjectParams projectParams, String templateIdentifier, String templateVersionLabel, PageParams pageParams) {
+    List<MonitoredService> monitoredServiceList =
+        hPersistence.createQuery(MonitoredService.class)
+            .disableValidation()
+            .filter(MonitoredServiceKeys.accountId, projectParams.getAccountIdentifier())
+            .filter(MonitoredServiceKeys.orgIdentifier, projectParams.getOrgIdentifier())
+            .filter(MonitoredServiceKeys.projectIdentifier, projectParams.getProjectIdentifier())
```
Comment (karansaraswat19) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java:2354]: these will be templateProjectParams, therefore orgIdentifier and projectIdentifier can be null also as templates can be account scoped as well

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java [hunk 2]
```diff
@@ -2340,6 +2343,45 @@ public List<ActiveServiceDTO> listActiveMonitoredServices(ProjectParams projectP
...
+            .filter(MonitoredServiceKeys.templateMetadata + "." + TemplateMetadataKeys.isTemplateByReference, true)
+            .filter(MonitoredServiceKeys.templateMetadata + "." + TemplateMetadataKeys.templateIdentifier,
+                templateIdentifier)
+            .filter(
+                MonitoredServiceKeys.templateMetadata + "." + TemplateMetadataKeys.versionLabel, templateVersionLabel)
+            .asList();
+
+    String templateInput = templateFacade.getTemplateInputs(projectParams, templateIdentifier, templateVersionLabel);
```
Comment (karansaraswat19) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java:2362]: can we name it as templateInputs

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/resources/MonitoredServiceResource.java [hunk 3]
```diff
@@ -779,4 +780,18 @@ public RestResponse<List<MonitoredServiceChangeDetailSLO>> getMonitoredServiceCh
     return ResponseDTO.newResponse(monitoredServiceService.getResolvedTemplateInputs(
         scopedProjectParams.getProjectParams(), identifier, templateIdentifier, versionLabel));
   }
+
+  @GET
+  @Path("/reconciliation-status")
+  @Timed
+  @ExceptionMetered
+  @ResponseMetered
+  @NGAccessControlCheck(resourceType = MONITORED_SERVICE, permission = EDIT_PERMISSION)
```
Comment (karansaraswat19) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/resources/MonitoredServiceResource.java:789]: do we need the EDIT permission here?

Comment (naveen-harness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/resources/MonitoredServiceResource.java:789]: I don't think so, thanks for pointing out, can we fix in the next PR ?

**Comment by naveen-harness** [comment]
trigger UTs

**Comment by naveen-harness** [comment]
trigger uts

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/beans/monitoredService/MonitoredServiceReference.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+@Data
+@Builder
+@AllArgsConstructor
+public class MonitoredServiceReference {
+  @NotNull String orgIdentifier;
+  @NotNull String projectIdentifier;
+  @NotNull String accountIdentifier;
+  String templateIdentifier;
+  String serviceIdentifier;
```
Comment (Varsha-lalwani) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/beans/monitoredService/MonitoredServiceReference.java:23]: Do we not need monitoredServiceIdentifier here?

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java [hunk 5]
```diff
@@ -2362,6 +2365,46 @@ public List<ActiveServiceDTO> listActiveMonitoredServices(ProjectParams projectP
     return activeServiceDTOList;
   }
 
+  @Override
+  public PageResponse<MonitoredServiceReference> getMonitoredServiceReconciliationStatuses(
+      ProjectParams projectParams, String templateIdentifier, String templateVersionLabel, PageParams pageParams) {
+    List<MonitoredService> monitoredServiceList =
+        hPersistence.createQuery(MonitoredService.class)
+            .disableValidation()
+            .filter(MonitoredServiceKeys.accountId, projectParams.getAccountIdentifier())
```
Comment (Varsha-lalwani) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java:2374]: what about template in different scope than monitored service?

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java [hunk 6]
```diff
@@ -2362,6 +2365,46 @@ public List<ActiveServiceDTO> listActiveMonitoredServices(ProjectParams projectP
...
+            .filter(MonitoredServiceKeys.orgIdentifier, projectParams.getOrgIdentifier())
+            .filter(MonitoredServiceKeys.projectIdentifier, projectParams.getProjectIdentifier())
+            .filter(MonitoredServiceKeys.templateMetadata + "." + TemplateMetadataKeys.isTemplateByReference, true)
+            .filter(MonitoredServiceKeys.templateMetadata + "." + TemplateMetadataKeys.templateIdentifier,
+                templateIdentifier)
+            .filter(
```
Comment (Varsha-lalwani) [inline srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java:2380]: version shouldn't be in filter, no?

**Review by Varsha-lalwani — CHANGES_REQUESTED** [review]

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java (hunk 5 above)
Comment (naveen-harness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java:2374]: template and monitored service have to be in the same scope

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/beans/monitoredService/MonitoredServiceReference.java (hunk 4 above)
Comment (naveen-harness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/beans/monitoredService/MonitoredServiceReference.java:23]: I think we should add, will add in the next PR

File: srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java (hunk 6 above)
Comment (naveen-harness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java:2380]: versionLabel needs to be in the filter, we're filtering by templateIdentifier and version

Comment (Varsha-lalwani) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java:2381]: we need to filter irrespective of template version right? template version will be used for reconcilation status only.

Comment (naveen-harness) [reply srm-service/modules/cv-nextgen-service/service/src/main/java/io/harness/cvng/core/services/impl/monitoredService/MonitoredServiceServiceImpl.java:2381]: there are two things: versionLabel and versionNumber both are different things
versionLabel is external attribute
versionNumber is something internal for template which is incremented every time template is updated

**Comment by naveen-harness** [comment]
trigger uts

**Comment by karansaraswat19** [comment]
trigger ut6, ut9

**Review by deepak-harness — APPROVED** [review]
> Reviewed 3 of 8 files at r2, 4 of 6 files at r3, 2 of 3 files at r4, 1 of 1 files at r5, 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/55762)* status: 1 change requests. 4. unresolved discussions.

**Review by karansaraswat19 — APPROVED** [review]

**Review by Varsha-lalwani — APPROVED** [review]

**Pull request merged**